### ISSUE TITLE
Keras connectivity metadata is not set correctly in TF 2.2.0rc0

### ISSUE THREAD

**Issue by drasmuss** [tensorflow/tensorflow#37548]
**System information** 
- Have I written custom code (as opposed to using a stock
example script provided in TensorFlow): yes
- OS Platform and Distribution (e.g.,
Linux Ubuntu 16.04): Windows 10
- Mobile device (e.g. iPhone 8, Pixel 2, Samsung Galaxy) if
the issue happens on mobile device: N/A
- TensorFlow installed from (source or
binary): binary
- TensorFlow version (use command below): 2.2.0rc0
- Python version: 3.6.10
- Bazel
version (if compiling from source): N/A
- GCC/Compiler version (if compiling from
source): N/A
- CUDA/cuDNN version: N/A
- GPU model and memory: N/A

**Describe the current behavior**

If a Keras model is applied to inputs with mismatched (but compatible) shape, the model is applied correctly but none of the Keras connectivity metadata (e.g. `inbound/outbound_nodes` or `keras_history`) is updated.

**Describe the expected behavior**

The Keras connectivity metadata should be updated, so after the model is applied there should be new inbound/outbound nodes for that application, and all the Tensors created by that application should have their keras_history set appropriately (this is the behaviour in TF<=2.1.0).

**Standalone code to reproduce the issue** 

``` python
import tensorflow as tf


for input_shape in [(1,), (1, 1)]:
    print("input_shape", input_shape)

    sub_in = tf.keras.Input((1,))
    relu_layer = tf.keras.layers.ReLU()
    sub_out = relu_layer(sub_in)
    submodel = tf.keras.Model(sub_in, sub_out)

    assert len(relu_layer.inbound_nodes) == 1

    inp = tf.keras.Input(input_shape)
    out = submodel(inp)

    assert len(relu_layer.inbound_nodes) == 2
```

The `assert len(relu_layer.inbound_nodes) == 2` condition fails when `input_shape=(1, 1)`, indicating that no new inbound nodes were created when `submodel` was applied to `inp`.

**Other info / logs** Include any logs or source code that would be helpful to
diagnose the problem. If including tracebacks, please include the full
traceback. Large logs and files should be attached.

The issue is that when this function is applied https://github.com/tensorflow/tensorflow/blob/v2.2.0-rc0/tensorflow/python/keras/engine/network.py#L926 it creates new tensors, which do not have a keras_history set. Then this condition evaluates to `False` https://github.com/tensorflow/tensorflow/blob/v2.2.0-rc0/tensorflow/python/keras/engine/base_layer.py#L947, so the `set_connectivity_metadata` function is never called here https://github.com/tensorflow/tensorflow/blob/v2.2.0-rc0/tensorflow/python/keras/engine/base_layer.py#L952.

**Comment by gadagashwini** [comment]
Was able to replicate the issue with Tf 2.2.rc0.
Please find the gist [here](https://colab.sandbox.google.com/gist/gadagashwini/8b75a84496af15d6c8e7a70fe383c98e/untitled448.ipynb). Thanks!

**Comment by ashutosh1919** [comment]
I think gist created by @gadagashwini has some problem. 
Here is the gist created by @gadagashwini , 
```python
import tensorflow as tf
for input_shape in [(1,), (1, 1)]:
    print("input_shape", input_shape)
    
    # Code Section 1
    sub_in = tf.keras.Input((1,))
    relu_layer = tf.keras.layers.ReLU()
    sub_out = relu_layer(sub_in)
    submodel = tf.keras.Model(sub_in, sub_out)

    assert len(relu_layer.inbound_nodes) == 1

    # Code section 2
    inp = tf.keras.Input(input_shape)
    out = submodel(inp)

    assert len(relu_layer.inbound_nodes) == 2
```
Now, Take a look at section code 1 & 2. Observe that there is no use of `input_shape` in code section 1. And `relu_layer` is defined in code section 1. So, `len(relu.inbound_nodes)` is independent of the value of `input_shape`. And moreover, in code section 2 assert statement, you are asserting `inbound_nodes` of `relu_layer` which is not even used in `inp` and `out` in code section 2 and is independent of value of `input_shape`.
That's why it gives assertion error. I have verified the code. Below code works fine.
```python
import tensorflow as tf
for input_shape in [(1,), (1, 1)]:
    print("input_shape", input_shape)
    
    # Code Section 1
    sub_in = tf.keras.Input((1,))
    relu_layer = tf.keras.layers.ReLU()
    sub_out = relu_layer(sub_in)
    submodel = tf.keras.Model(sub_in, sub_out)

    assert len(relu_layer.inbound_nodes) == 1

    # Code section 2
    inp = tf.keras.Input(input_shape)
    out = relu_layer(inp)

    assert len(relu_layer.inbound_nodes) == 2
```
I have just added `relu_layer` instead of `submodel` in code section 2. 

@drasmuss , Please give your comments too.

**Comment by drasmuss** [comment]
> Observe that there is no use of input_shape in code section 1. And relu_layer is defined in code section 1. So, len(relu.inbound_nodes) is independent of the value of input_shape. And moreover, in code section 2 assert statement, you are asserting inbound_nodes of relu_layer which is not even used in inp and out in code section 2 and is independent of value of input_shape.

"Code section 1" creates `submodel`, which is used in "Code section 2". So effectively all of "Code section 1" is being repeated in "Code section 2". You can read more about how this works in Keras here https://www.tensorflow.org/guide/keras/functional#all_models_are_callable_just_like_layers.

> That's why it gives assertion error. I have verified the code. Below code works fine

Your code avoids the problem, it doesn't solve it (because the root of the issue is in this section https://github.com/tensorflow/tensorflow/blob/v2.2.0-rc0/tensorflow/python/keras/engine/network.py#L926, as pointed out above, which is not called if you're not using a submodel). The original example should work if the Keras connectivity were being set correctly, which you can verify by setting the TensorFlow version to any value <2.2.

**Comment by pavithrasv** [comment]
@drasmuss We are going to be updating the code to remove metadata from all nested layers going forward. 'inbound_nodes' property on layers has been deprecated as it is an internal implementation detail.

**Comment by drasmuss** [comment]
@pavithrasv Is there an alternate way to determine, given a tensor, what the internal layer was that generated that tensor? I.e., given the code example from above
```
    sub_in = tf.keras.Input((1,))
    relu_layer = tf.keras.layers.ReLU()
    sub_out = relu_layer(sub_in)
    submodel = tf.keras.Model(sub_in, sub_out)

    inp = tf.keras.Input(input_shape)
    out = submodel(inp)
```
when given `out`, I need to be able to retrieve `relu_layer`. The only way I was able to do this previously was by traversing through the metadata, but if that is being removed then I'll need another method.

**Comment by pavithrasv** [comment]
@drasmuss What is the use case for this? We do not currently have a method as this is internal implementation detail of functional model construction.

**Comment by drasmuss** [comment]
I need to be able to determine the graph structure of a functional Keras model (i.e. the full graph of which layers are connected to which other layers).

**Comment by pavithrasv** [comment]
Can you use `model.layers` to navigate and find the graph of layers in the inner model?

**Comment by drasmuss** [comment]
No, because `model.layers` just tells you which layer objects are in the model, not how they are connected to each other. For example, consider a model like this
```
inp = x = tf.keras.layers.Input(shape=(4,))
my_layer = tf.keras.layers.Dense(units=2)
x = my_layer(x)
x = my_layer(x)
x = my_layer(x)
model = tf.keras.Model(inp, x)
```
`model.layers` just shows one Input layer and one Dense layer. If you want to recreate the structure of that model, you have to look at the Keras connectivity metadata of `my_layer`.

**Comment by pavithrasv** [comment]
We do not have any public properties or methods for this today. We have `tensor._keras_history` which will provide that information, but since this is a private property we do not recommend relying on it (similar as `_inbound_nodes`).

**Comment by drasmuss** [comment]
I would love to not rely on private properties, but as you say there is no way to get this information through public properties. So if we want the Keras graph structure (which doesn't seem like a really unusual use case) that's our only option :smile:.

Do you have a timeline for when you plan on removing this functionality? And are there any plans to provide equivalent information through some other data structure, or will it just be impossible to recover the Keras graph structure at that point?

**Comment by pavithrasv** [comment]
I have fixed the original issue. ie. you can use `_keras_history` or `inbound_nodes` to get the information. There are no plans at this time to add properties describing the Keras graph structure.

**Comment by goldiegadde** [comment]
@drasmuss this issue has been fixed in the latest tf-nightly and the next TF 2.3.0 release should have it. Please feel free to reopen if it does not work for you.

**Comment by drasmuss** [comment]
Confirmed that this is fixed in TF 2.3.0rc0, thanks!
